tools/proto_format/proto_sync.py

```diff
@@ -1,109 +1,151 @@
 #!/usr/bin/env python3
 
-# Diff or copy pretty-printed artifacts to the source tree.
+# Diff or copy normalized artifacts to the source tree.
 
 import argparse
+import difflib
+import io
 import os
 import pathlib
-import shutil
-import subprocess
 import sys
 import tarfile
-import tempfile
-
-
-def git_status(path):
-    return subprocess.check_output(
-        ['git', 'status', '--porcelain', str(path)], cwd=str(path)).decode()
-
-
-def generate_current_api_dir(api_dir, dst_dir):
-    """Helper function to generate original API repository to be compared with diff.
-    This copies the original API repository and deletes file we don't want to compare.
-    Args:
-        api_dir: the original api directory
-        dst_dir: the api directory to be compared in temporary directory
-    """
-    contrib_dst = dst_dir.joinpath("contrib")
-    shutil.copytree(str(api_dir.joinpath("contrib")), str(contrib_dst))
-
-    dst = dst_dir.joinpath("envoy")
-    shutil.copytree(str(api_dir.joinpath("envoy")), str(dst))
-
-    # envoy.service.auth.v2alpha exist for compatibility while we don't run in protoxform
-    # so we ignore it here.
-    shutil.rmtree(str(dst.joinpath("service", "auth", "v2alpha")))
-
-    for p in dst.glob('**/*.md'):
-        p.unlink()
-
-
-def sync(api_root, formatted, mode, is_ci):
-    api_root_path = pathlib.Path(api_root)
-    with tempfile.TemporaryDirectory() as tmp:
-        tmp_path = pathlib.Path(tmp)
-
-        # a - the actual api directory found on fs
-        current_api_dir = tmp_path.joinpath("a")
-        current_api_dir.mkdir(0o755, True, True)
-        api_root_path = pathlib.Path(api_root).absolute()
-        generate_current_api_dir(api_root_path, current_api_dir)
-
-        # b - ideally formatted version of api directory according to bazel
-        dst_dir = tmp_path.joinpath("b")
-        with tarfile.open(formatted) as tar:
-            tar.extractall(dst_dir)
-
-        # These support files are handled manually.
-        for f in ['envoy/annotations/resource.proto', 'envoy/annotations/deprecation.proto',
-                  'envoy/annotations/BUILD']:
-            copy_dst_dir = pathlib.Path(dst_dir, os.path.dirname(f))
-            copy_dst_dir.mkdir(exist_ok=True, parents=True)
-            shutil.copy(str(pathlib.Path(api_root, f)), str(copy_dst_dir))
-
-        diff = subprocess.run(['diff', '-Npur', "a", "b"], cwd=tmp, stdout=subprocess.PIPE).stdout
-
-        if diff.strip():
+from hashlib import sha1
+
+# TODO(phlax): Move all of this code to `envoy.code.check`
+
+
+class Git(object):
```

Comment:
by way of explanation - this is here to avoid a dependency on `gitpython` or similar and to avoid the more expensive way of creating a python git repo and deriving the git hash/headers/diff from that

Reply:
its not just the dependency - it would add as much code and is less efficient - as discussed offline